superduperdb/vector_search/lance.py

```diff
@@ -0,0 +1,178 @@
+        uri: str,
+        dimensions: int,
+    ) -> None:
+        """
+        :param uri: URI of the Lance dataset
+        :param dimensions: dimension of the vector embeddings in the Lance dataset
+        """
+        self.uri = uri
+        self.dimensions = dimensions
+
+    @property
+    def _dataset(self) -> lance.LanceDataset:
+        try:
+            return lance.dataset(self.uri)
+        except Exception as e:
+            raise Exception(f'Could not load Lance dataset at {self.uri}') from e
+
+    # TODO: ``upsert`` kwarg is not included in the ABC interface but it is assumed part
+    # of the interface in ``superduperdb/db/mongodb/cdc/vector_task_factory.py``!
+    # Short-term fix is to include it here, long-term fix is to tidy up ABC interface.
+    def add(self, data: t.Sequence[VectorCollectionItem], upsert: bool = False) -> None:
+        """
+        Add vectors to existing Lance dataset or create a new one if it doesn't exist.
+
+        :param data: t.Sequence of ``VectorCollectionItem`` objects.
+        """
+
+        typ = pa.list_(
+            pa.field('values', pa.float32(), nullable=False), self.dimensions
+        )
+        _vecs = pa.array([item.vector for item in data], type=typ)
+        _ids = pa.array([item.id for item in data], type=pa.string())
+        _table = pa.Table.from_arrays([_ids, _vecs], names=['id', 'vector'])
+
+        if not os.path.exists(self.uri):
+            lance.write_dataset(_table, self.uri, mode='create')
+        else:
+            lance.write_dataset(_table, self.uri, mode='append')
+
+        # TODO: Add logic to optimise Lance dataset for vector search at
+        # regular intervals or after a certain number of items have been added.
+        # These optimisations include fragment merging, compacting, deleting, etc.
+        # See ``lance.DatasetOptimizer`` for further details.
+
+    def find_nearest_from_id(
+        self,
+        identifier: VectorCollectionItemId,
+        *,
+        within_ids: t.Sequence[VectorCollectionItemId] = (),
+        limit: int = 100,
+        offset: int = 0,
+    ) -> t.List[VectorCollectionResult]:
+        """
+        Find items that are nearest to the item with the given identifier.
+
+        The ``lance`` file format has been specifically designed for fast
+        random access. The logic to take advantage of this is implemented
+        by the ``.take`` method.
+
+        See https://blog.lancedb.com/benchmarking-random-access-in-lance-ed690757a826
+        for further details.
+
+        :param identifier: identifier of the item
+        :param within_ids: identifiers to search within
+        :param limit: maximum number of nearest items to return
+        :param offset: offset of the first item to return
+        """
+
+        # ``.take`` expect a list of integers or pyarrow array as input
+        vector = self._dataset.take([int(identifier)], columns=['vector']).to_pydict()[
+            'vector'
+        ][0]
```

Comment:
how can i define type of search
cosine, l1, l2 etc

Reply:
can we create a function like `def get` which return vector given an id

Reply:
`get` isn't part of the base class interface. Why do we need it (we shouldn't rely on it if it isn't part of the interface)?

Reply:
`measure` isn't part of the signature for the base class. It is only part of the signature for `VectorIndex` (which I haven't implemented yet).